Fix error in bin size for `sensitivity.outcome_map`

Rosejoycrocker opened the pull request:
Fixes bug where the number of bins was being set by the maximum number of categories where a categorical variable was used. Also moves similar code used in `sensitivity.rsa` into functions to be used by both methods.

Closes #615

ConnectedSystems commented on ext/AvizExt/viz/sensitivity.jl:
```diff
@@ -2,6 +2,10 @@ using Statistics
 using Printf
 using ADRIA.sensitivity: _get_cat_quantile
 
+const FV_LABELS = [
+    "cf", "unguided", last.(split.(string.(ADRIA.decision.mcda_methods()), "."))...
+]
```
Suggest changing this to a function that returns the string vector to avoid having to restart Julia if the constant needs to be changed.

ConnectedSystems commented on src/analysis/sensitivity.jl:
```diff
@@ -82,6 +82,72 @@ function _get_cat_quantile(
+    return seq_store
+end
+
+function _create_yax_tuple_store(
+    seq_store::Dict{Symbol,Vector{Float64}}, foi_spec::DataFrame,
+    unordered_cat::Vector{Symbol}; second_dim::Union{Dim,Vector{Any}}=[],
+)
```
Add docstrings to this please, it's more than a few lines and it seems important.

ConnectedSystems commented on ext/AvizExt/viz/sensitivity.jl:
```diff
@@ -308,7 +310,7 @@ Makie figure
     g::Union{GridLayout,GridPosition},
     rs::ResultSet,
-    outcomes::NamedDimsArray,
+    outcomes::Dataset,
```
Can we make sure this works with a YAXArray as well?

e.g., we just want to plot results for single factor.

Rosejoycrocker replied:
I'll make this work

ConnectedSystems reviewed (changes requested):
Some questions/comments for your consideration

Rosejoycrocker commented on src/analysis/sensitivity.jl:
```diff
@@ -82,6 +82,72 @@ function _get_cat_quantile(
+    )
+    r_s_cat = NamedTuple(zip(Tuple(unordered_cat), yax_store_cat))
+    r_s = merge(r_s_cat, r_s_default)
+    return r_s
```
Yeah we could do that in `_create_yax_tuple_store` but I think you'd still have to create the NamedTuple to get the optional argument names in `Dataset()`

Rosejoycrocker commented on src/analysis/sensitivity.jl:
```diff
@@ -82,6 +82,72 @@ function _get_cat_quantile(
+    # Create storage NamedTuples for unordered categorical variables and other variables, then merge
+    r_s_default = NamedTuple(
+        zip(
+            Tuple(foi_spec.fieldname[foi_spec.ptype .!= "unordered categorical"]),
```
Yes, but these don't need different S, they just need to use `_get_cat_quantile()` to calculate their quantiles (which I've added further down). This means for variables like `N_seed_TA` we don't have thousands of categories we just have 10 but these categories are discrete.

ConnectedSystems commented on ext/AvizExt/viz/sensitivity.jl:
```diff
@@ -1,6 +1,18 @@
+"""
+    _get_guided_labels()::Vector{String}
+
+    Returns labels for categories of the `guided` factor.
```
No indentation for the description.

```suggestion
Returns labels for categories of the `guided` factor.
```

ConnectedSystems commented on src/analysis/sensitivity.jl:
```diff
@@ -82,6 +83,116 @@ function _get_cat_quantile(
 end
 
+"""
+    _get_factor_quantile(seq_store::Dict{Symbol,Vector{Float64}},foi_spec::DataFrame,
+        fact_t::Symbol)
```
Spacing?

ConnectedSystems commented on test/runtests.jl:
```diff
@@ -338,29 +338,34 @@ function test_rs_w_fig()
+    )
+
+    # Indicate factor values that are in the top 30 percentile
+    tac_om_70 = ADRIA.sensitivity.outcome_map(
+        rs, mean_s_tac, x -> any(x .>= 0.7), foi; S=20
+    )
```
Minor comment, but the variable should be `tac_om_30` to represent >= 30th percentile

ConnectedSystems replied:
Yes, I know this was an oversight from the previous implementation

ConnectedSystems commented on ext/AvizExt/viz/sensitivity.jl:
```diff
@@ -429,18 +414,86 @@ function ADRIA.viz.outcome_map!(
 
     return f
 end
+function ADRIA.viz.outcome_map!(
```
Docstrings aren't updated

ConnectedSystems replied:
Docstrings still not updated?

ConnectedSystems commented on src/analysis/sensitivity.jl:
```diff
@@ -82,6 +83,116 @@ function _get_cat_quantile(
+- `foi_spec` : Model specification for factors of interest
+- `X_f` : Scenario dataframe for factor of interest
+- `factor_name` : Contains true where the factor is categorical and false otherwise
+"""
```
What does this return?

ConnectedSystems commented on src/analysis/sensitivity.jl:
```diff
@@ -82,6 +83,116 @@ function _get_cat_quantile(
+    S::Int64)::Dict{Symbol,Vector{Float64}}
+    seq_store::Dict{Symbol,Vector{Float64}} = Dict() # storage for bin sequences
+
+    # Get unique bin sequences for unordered categorical variables and store
+    for factor in unordered_cat
+        S_temp = _category_bins(model_spec[model_spec.fieldname .== factor, :])
+        seq_store[factor] = collect(0.0:(1 / S_temp):1.0)
+    end
+    # Other variables have default sequence using input S
+    seq_store[:default] = collect(0.0:(1 / S):1.0)
```
Fix spacing please

Rosejoycrocker commented:
I think I've addressed your comments, so ready for review again @ConnectedSystems .

Rosejoycrocker commented:
@ConnectedSystems and @Zapiano, this has been rebased against `yax_main` and should now address #713 and your previous comments.

Example current figures for `outcome_map`:

![outcome_map_ex](https://github.com/open-AIMS/ADRIA.jl/assets/56939532/5fec2e72-83d0-4028-bfa6-e6b53f23830d)

Example current figures for `rsa`:
![rsa_ex](https://github.com/open-AIMS/ADRIA.jl/assets/56939532/071a39b9-a098-4943-b0c5-54578e0100ac)

Rosejoycrocker commented:
This has been rebased with the most recent version of main now, @ConnectedSystems . I've tested the functions and `outcome_map` and `rsa` (plus visualisations) are still working as expected with the rebase.

ConnectedSystems commented:
@Rosejoycrocker the rebase you did (at least I think it was a rebase?) brought in some weird artifacts:

![image](https://github.com/open-AIMS/ADRIA.jl/assets/4075136/b08da5e4-8e96-4b1c-9706-b7220d5d945e)

`MatDomain.jl` is a file that has long since been removed and there seems to be a local copy of the repository (probably something done by a VS Code plugin or similar).

Are you able to clean these up?

Rosejoycrocker commented:
> @Rosejoycrocker the rebase you did (at least I think it was a rebase?) brought in some weird artifacts:
> 
> ![image](https://private-user-images.githubusercontent.com/4075136/316877826-b08da5e4-8e96-4b1c-9706-b7220d5d945e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE0OTE4NDEsIm5iZiI6MTcxMTQ5MTU0MSwicGF0aCI6Ii80MDc1MTM2LzMxNjg3NzgyNi1iMDhkYTVlNC04ZTk2LTRiMWMtOTcwNi1iNzIyMGQ1ZDk0NWUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjZUMjIxOTAxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZWE5ZmQ3MzIyNjA4ODI0NWNkYzk3OTdkYjUxYWIyZTI4Nzg0YWZhNmNiM2M4ZjEyNDI2YmZkMDcyY2Q2MTA2YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.bwHijHby2yuTP88eD_KAMakl6moY9mnK_eUK5rmuX3g)
> 
> `MatDomain.jl` is a file that has long since been removed and there seems to be a local copy of the repository (probably something done by a VS Code plugin or similar).
> 
> Are you able to clean these up?

This might be because this PR was originally rebased with main, then we switched to rebasing with yax_man, and now again with main. I'll see what I can do.

ConnectedSystems commented:
I've cleared out the duplicate commits and rebased onto `main`.

Last thing to do Rose, aside from the comment above: could you please provide the example snippet used to produce figures above as well as add the updated figures here?

Rosejoycrocker commented:
![outcome_map_ex](https://github.com/open-AIMS/ADRIA.jl/assets/56939532/7f5493fc-0784-4446-8318-6ebd4f68b505)
![rsa_ex](https://github.com/open-AIMS/ADRIA.jl/assets/56939532/aaf5182a-3bf4-4f17-a840-4057ad2dcf2f)

Here are example plots for `outcome_map` and `rsa`, plotted using:

```julia
si = ADRIA.sensitivity.outcome_map(
    rs, y, rule, [:dhw_scenario, :guided, :N_seed_TA, :N_seed_CA]
)

fig = ADRIA.viz.outcome_map(rs, si, [:dhw_scenario, :guided, :N_seed_TA, :N_seed_CA])

si = ADRIA.sensitivity.rsa(
    scens[:, [:dhw_scenario, :guided, :N_seed_TA, :N_seed_CA]], y, ADRIA.model_spec(rs)
)

fig = ADRIA.viz.rsa(rs, si, [:dhw_scenario, :guided, :N_seed_TA, :N_seed_CA])
```

Rosejoycrocker commented:
Hi @ConnectedSystems, could you give this a final (hopefully) review? I rebased with main again and checked that the plots are still plotting as in the example plots above, which they are.

Zapiano commented:
@Rosejoycrocker Can I make a suggestion? You have 33 commits changing 6 files. This is ~ 5 commits per  file - which is too much. Maybe squashing some commits would be helpful. :)

Rosejoycrocker commented:
> @Rosejoycrocker Can I make a suggestion? You have 33 commits changing 6 files. This is ~ 5 commits per file - which is too much. Maybe squashing some commits would be helpful. :)

Thanks, I've squashed into 9 commits to make it cleaner :)

ConnectedSystems commented:
It's not clear to me what `y` and `rule` is in your example but mocking those up and running it with just counterfactual scenarios, I get a plot like this:

![image](https://github.com/open-AIMS/ADRIA.jl/assets/4075136/c8fbfaad-ec53-4c61-bf2b-77379f48c237)

Note that the `unguided` tick label makes an appearance and the "seeded" factors seem to be non-zero.
Which is puzzling to me because they definitely are zero in the scenario spec.


Trying out the regional sensitivity analysis with a mix of counterfactual/guided/unguided results in a whole bunch of warning messages followed by an exception:

```julia
┌ Warning: `cat` cannot concatenate `Dimension`s, falling back to `parent` type:
│ `Ordered` lookups are misaligned at 384 and 1 on dimension DimensionalData.Dimensions.Dim{:scenarios}.
│ 
│ To fix for `AbstractDimArray`, pass new lookup values as `cat(As...; dims=DimensionalData.Dimensions.Dim{:scenarios}(newlookupvals))` keyword or `dims=DimensionalData.Dimensions.Dim{:scenarios}()` for empty `NoLookup`.
└ @ DimensionalData C:\Users\tiwanaga\.julia\packages\DimensionalData\BoJag\src\array\methods.jl:534

# ... several more lines like the above ...

ERROR: MethodError: no method matching HypothesisTests.KSampleADTest(::Int64, ::Int64, ::Float64, ::Float64, ::Bool, ::Int64, ::YAXArrays.Cubes.YAXArray{…}, ::Vector{…})

Closest candidates are:
  HypothesisTests.KSampleADTest(::Int64, ::Int64, ::Float64, ::Float64, ::Bool, ::Int64, ::Vector{T}, ::Vector{Int64}) where T<:Real
   @ HypothesisTests C:\Users\tiwanaga\.julia\packages\HypothesisTests\r322N\src\anderson_darling.jl:84
  HypothesisTests.KSampleADTest(::AbstractVector{T}...; modified, nsim) where T<:Real
   @ HypothesisTests C:\Users\tiwanaga\.julia\packages\HypothesisTests\r322N\src\anderson_darling.jl:117

Stacktrace:
 [1] a2_ksample(samples::Tuple{YAXArrays.Cubes.YAXArray{…}, YAXArrays.Cubes.YAXArray{…}}, modified::Bool, method::Int64)
   @ HypothesisTests C:\Users\tiwanaga\.julia\packages\HypothesisTests\r322N\src\anderson_darling.jl:310
 [2] HypothesisTests.KSampleADTest(::YAXArrays.Cubes.YAXArray{…}, ::Vararg{…}; modified::Bool, nsim::Int64)
   @ HypothesisTests C:\Users\tiwanaga\.julia\packages\HypothesisTests\r322N\src\anderson_darling.jl:117
 [3] HypothesisTests.KSampleADTest(::YAXArrays.Cubes.YAXArray{…}, ::Vararg{…})
   @ HypothesisTests C:\Users\tiwanaga\.julia\packages\HypothesisTests\r322N\src\anderson_darling.jl:117
 [4] rsa(r_s::YAXArrays.Cubes.YAXArray{…}, X_q::Vector{…}, X_i::Vector{…}, y::YAXArrays.Cubes.YAXArray{…}, sel::BitVector)
   @ ADRIA.sensitivity c:\Users\tiwanaga\projects\ADRIA.jl\src\analysis\sensitivity.jl:634
 [5] rsa(X::DataFrames.DataFrame, y::YAXArrays.Cubes.YAXArray{…}, model_spec::DataFrames.DataFrame; S::Int64)
   @ ADRIA.sensitivity c:\Users\tiwanaga\projects\ADRIA.jl\src\analysis\sensitivity.jl:619
 [6] rsa(X::DataFrames.DataFrame, y::YAXArrays.Cubes.YAXArray{…}, model_spec::DataFrames.DataFrame)
   @ ADRIA.sensitivity c:\Users\tiwanaga\projects\ADRIA.jl\src\analysis\sensitivity.jl:597
 [7] top-level scope
   @ REPL[44]:1
Some type information was truncated. Use `show(err)` to see complete types.
```

Here's the example I've mocked up:

```julia
n_samples = 128
cf_scens = ADRIA.sample_cf(dom, n_samples)
ug_scens = ADRIA.sample_unguided(dom, n_samples)
piv_scens = ADRIA.sample_guided(dom, n_samples)
scens = vcat(cf_scens, ug_scens, piv_scens)

rs26 = ADRIA.run_scenarios(dom, scens, "26")
src26 = ADRIA.metrics.scenario_relative_cover(rs26)

y = dropdims(mean(src26, dims=1), dims=1)
rule = x -> all(x .> 0.5)

foi = [:dhw_scenario, :guided, :N_seed_TA, :N_seed_CA]
out = ADRIA.sensitivity.outcome_map(
    rs26, y, rule, foi
)

fig = ADRIA.viz.outcome_map(rs26, out, foi)


si = ADRIA.sensitivity.rsa(
    scens[:, foi], y, ADRIA.model_spec(rs26)
)

fig = ADRIA.viz.rsa(rs26, si, foi)

```

ConnectedSystems commented on ext/AvizExt/viz/sensitivity.jl:
```diff
@@ -429,18 +482,103 @@ function ADRIA.viz.outcome_map!(
+        title_val = ms_factor.name[1]
+    end
+
+    ax::Axis = Axis(g[1, 1]; title=title_val,
+        titlesize=38,
+        xlabel=xlabel,
+        ylabel=ylabel,
+        xlabelsize=32,
+        ylabelsize=32,
+        ylabelrotation=π / 2.0, axis_opts...)
```
Formatting?

Rosejoycrocker commented on ext/AvizExt/viz/sensitivity.jl:
```diff
@@ -349,43 +425,20 @@ function ADRIA.viz.outcome_map!(
-                fv_s[.!ismissing.(outcomes[factors=At(f_name), CI=At(:lower)])],
-                collect(skipmissing(outcomes[factors=At(f_name), CI=At(:lower)])),
-                collect(skipmissing(outcomes[factors=At(f_name), CI=At(:upper)])),
+                g[r, c]; title=ms_factor.name[1], axis_opts...
```
No, it's the human readable name rather than the indexing symbol name

Rosejoycrocker commented:
Sorry @ConnectedSystems , I thought I'd included `y` and `rule` but I hadn't, they are:

```julia
s_tac = ADRIA.metrics.scenario_total_cover(rs)
y = vec(mean(s_tac; dims=1))
rule = y -> all(y .> 0.5)
```

I stuck with y being an `AbstractVector` as in the original, I thought this might be so that the function was more generalisable, but I've now changed it to a `YAXArray{Float64,1}` in the wrapper functions, so in the new implementation,

```julia
s_tac = ADRIA.metrics.scenario_total_cover(rs)
y = dropdims(mean(s_tac; dims=1); dims=1)
rule = y -> all(y .> 0.5)

si = ADRIA.sensitivity.outcome_map(
    rs, y, rule, [:dhw_scenario, :guided, :N_seed_TA, :N_seed_CA]
)

fig = ADRIA.viz.outcome_map(rs, si, [:dhw_scenario, :guided, :N_seed_TA, :N_seed_CA])

si = ADRIA.sensitivity.rsa(rs, y)

fig = ADRIA.viz.rsa(rs, si, [:dhw_scenario, :guided, :N_seed_TA, :N_seed_CA])
```

which works for me. I also noticed `y` isn't defined in the analysis.md example either so I've added it.

On the counterfactuals only example, I didn't try plotting intervention parameters when only cfs are run as I didn't think this would be done, but I'll look into it now.

ConnectedSystems commented:
> On the counterfactuals only example, I didn't try plotting intervention parameters when only cfs are run as I didn't think this would be done, but I'll look into it now.

To clarify the intent there, it's to pick up issues when working with constant or otherwise inactive factors. It's not strictly related to doing analyses on counterfactual scenarios.

Rosejoycrocker commented:
> > On the counterfactuals only example, I didn't try plotting intervention parameters when only cfs are run as I didn't think this would be done, but I'll look into it now.
> 
> To clarify the intent there, it's to pick up issues when working with constant or otherwise inactive factors. It's not strictly related to doing analyses on counterfactual scenarios.

Thanks for clarifyin. I think what is in your plot occurs because all values are set as `missing`, so maybe I'll put in something to skip the plot if this occurs

ConnectedSystems commented:
> maybe I'll put in something to skip the plot if this occurs

A blank plot would be better. Hiding a plot will cause more confusion.

Rosejoycrocker commented:
> > maybe I'll put in something to skip the plot if this occurs
> 
> A blank plot would be better. A missing plot will cause more confusion.

Ok thanks, will do

Rosejoycrocker commented:
I've fixed formatting issues @ConnectedSystems, so I think this is ready for merging now unless you have any final comments.

ConnectedSystems commented:
There are still some formatting issues and I see newer code is overwritten.

I'll look over it overnight and fix it up.

Rosejoycrocker commented:
> There are still some formatting issues and I see newer code is overwritten.
> 
> I'll look over it overnight and fix it up.

I tested the functions and plotting are working post rebase. I'll doule check my formatting file is up to date but it seemed to pass the format check on my end

ConnectedSystems commented on src/analysis/sensitivity.jl:
```diff
@@ -404,8 +534,10 @@ end
-    rsa(rs::ResultSet, y::AbstractVector{<:Real}; S::Int64=10)::Dataset
-    rsa(rs::ResultSet, y::AbstractArray{<:Real}, factors::Vector{Symbol}; S::Int64=10)::Dataset
+    rsa(X::DataFrame, y::Vector{<:Real}, factor::Symbol, model_spec::DataFrame; S::Int64=10)::YAXArray
+    rsa(rs::ResultSet, y::AXArray{Float64,1}; S::Int64=10)::Dataset
+    rsa(rs::ResultSet, y::AXArray{Float64,1}, factors::Vector{Symbol}; S::Int64=10)::Dataset
+    rsa(rs::ResultSet, y::AXArray{Float64,1}, factor::Symbol; S::Int64=10)::YAXArray
```
FYI the formatting check doesn't cover issues like this

ConnectedSystems commented on ext/AvizExt/viz/sensitivity.jl:
```diff
@@ -437,18 +489,107 @@ function ADRIA.viz.outcome_map!(
+    rs::ResultSet,
+    outcomes::YAXArray,
+    factor::Symbol;
+    opts::Dict=Dict(),
+    fig_opts::Dict=Dict(),
+    axis_opts::Dict=Dict()
```
This set of changes is replacing the new approach to defining default options for visualizations (see #777).

I'm assuming the change is being accidentally overwritten.

Rosejoycrocker commented:
> > There are still some formatting issues and I see newer code is overwritten.
> > I'll look over it overnight and fix it up.
> 
> I tested the functions and plotting are working post rebase. I'll doule check my formatting file is up to date but it seemed to pass the format check on my end

Yeah sorry, must have happened in the rebase, I'll fix this

ConnectedSystems commented:
Leave it with me please @Rosejoycrocker , I'll get to this eventually

Rosejoycrocker commented:
> Leave it with me please @Rosejoycrocker , I'll get to this eventually

Thanks, I'll leave it then

ConnectedSystems reviewed (approved):
Merged in fixes, confirming everything appears to work as expected, with the exception of tests. We can fix those if they error later.
